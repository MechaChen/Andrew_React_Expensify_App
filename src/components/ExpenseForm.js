import React from "react";
import { SingleDatePicker } from "react-dates";
import moment from "moment";
import "react-dates/lib/css/_datepicker.css";

// setup note state
// setup onChange and value for textarea

export default class ExpenseForm extends React.Component {
  state = {
    description: this.props.expense.description,
    note: this.props.expense.note,
    amount: (this.props.expense.amount / 100).toString(),
    createdAt: moment(this.props.expense.createdAt),
    calendarFocused: false,
    error: null
  };
  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  onAmountChange = e => {
    const amount = e.target.value;
    if (!amount || /^\d+(\.\d{0,2})?$/.test(amount)) {
      this.setState(() => ({ amount }));
    }
  };
  onNoteChange = e => {
    e.persist();
    this.setState(() => ({ note: e.target.value }));
  };
  onDateChange = createdAt => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };
  onSubmit = e => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      // Set error state equal to 'Please provide description and amount.'
      this.setState(() => ({
        error: "Please provide description and amount."
      }));
    } else {
      // Clear the error
      this.setState(() => ({ error: null }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };
  render() {
    return (
      <div>
        <form onSubmit={this.onSubmit}>
          {this.state.error && <p>{this.state.error}</p>}
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder="Add a note for your expense(optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
          ></textarea>
          <button>Add Expense</button>
        </form>
      </div>
    );
  }
}

ExpenseForm.defaultProps = {
  description: "",
  amount: "",
  note: "",
  createdAt: moment()
};
