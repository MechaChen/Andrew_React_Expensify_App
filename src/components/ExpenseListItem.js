// Export a stateless functional Component
// decription, amount, createdAt

import React from "react";

const ExpenseListItem = props => (
  <div>
    <p>Description: {props.description}</p>
    <p>Amount: {props.amount}</p>
    <p>Created At: {props.createdAt}</p>
  </div>
);

export default ExpenseListItem;
