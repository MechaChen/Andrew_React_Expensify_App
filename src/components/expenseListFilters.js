import React from "react";
import { connect } from "react-redux";
import { setTextFilter } from "../actions/filters";

const ExpenseListFilters = props => (
  <div>
    <input
      type="text"
      value={props.text}
      onChange={e => {
        props.dispatch(setTextFilter(e.target.value));
      }}
    />
  </div>
);

const mapStateToProps = state => ({
  text: state.filters.text
});

export default connect(mapStateToProps)(ExpenseListFilters);
