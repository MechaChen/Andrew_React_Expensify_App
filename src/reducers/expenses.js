// Expense Reducer

const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case "ADD_EXPENSE":
      return [...state, action.expense];
    case "EDIT_EXPENSE":
      return state.map(expense =>
        expense.id === action.id ? { ...expense, ...action.updates } : expense
      );
    case "REMOVE_EXPENSE":
      return state.filter(({ id }) => id !== action.id);
    default:
      return state;
  }
};
