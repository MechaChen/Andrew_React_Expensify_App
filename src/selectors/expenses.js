// Get visible expenses
export default getVisibleExpenses = (
  expenses,
  { text, sortBy, startDate, endDate }
) => {
  return expenses
    .filter(expense => {
      const startDateMatch =
        typeof startDate !== "number" || expense.createdAt >= startDate;
      const endDateMatch =
        typeof endDate !== "number" || expense.createdAt <= endDate;

      // figure out if expenses.description has the text variable string inside of it
      // includes
      // convert both strings to lower case
      const textMatch =
        typeof text !== "string" ||
        expense.description.toLowerCase().includes(text.toLowerCase());

      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      if (sortBy === "date") {
        return a.createdAt < b.createdAt ? 1 : -1;
      } else if (sortBy === "amount") {
        return a.amount < b.amount ? 1 : -1;
      }

      // sortBy -> amount
      // put the ones with a greater amount first
    });
};
