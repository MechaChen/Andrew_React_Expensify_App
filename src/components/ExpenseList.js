import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpense from "../selectors/expenses";

const ExpenseList = props => (
  <div>
    <h1>Expense List</h1>
    {props.expenses.map(expense => (
      <ExpenseListItem key={expense.id} {...expense} />
    ))}
  </div>
);

// connect(1)(2) : like the function we created on 100 tutorial
//  1. a function return the data we need from store
//  2. Inner Component

const mapStateToProps = state => ({
  expenses: selectExpense(state.expenses, state.filters)
});

export default connect(mapStateToProps)(ExpenseList);

// export default ConnectedExpenseList;
