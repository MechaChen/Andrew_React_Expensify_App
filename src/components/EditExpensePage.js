import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { editExpense, removeExpense } from "../actions/expenses";

// Refactor EditExpensePage to be a class based component
// Setup mapDispatchToProp editExpense and removeExpense
export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  }
  onRemove = () => {
    this.props.removeExpense(this.props.expense.id);
    this.props.history.push('/');
  }
  render() {
    return (
      <div>
      <ExpenseForm
        expense={this.props.expense}
        onSubmit={this.onSubmit}
      />
      <button
        onClick={this.onRemove}
      >
        Remove
      </button>
    </div>
    );
  }
}

// should render EditExpensePage
// snapshot

// should handle editExpense
// spies

// should handle removeExpense
// spies


const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch) => ({
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  removeExpense: (id) => dispatch(removeExpense(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
