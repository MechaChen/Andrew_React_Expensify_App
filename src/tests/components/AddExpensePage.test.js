import React from 'react';
import { shallow } from 'enzyme';
import { AddExpensePage } from '../../components/AddExpensePage';
import expenses from '../fixtures/expenses';

test('should render AddExpesePage', () => {
  const onSubmit = jest.fn();
  const history = { push: jest.fn() };
  const wrapper = shallow(
    <AddExpensePage onSubmit={onSubmit} history={history} />
  );
  expect(wrapper).toMatchSnapshot();
});

test('should handle onSubmit', () => {
  const onSubmit = jest.fn();
  const history = { push: jest.fn() };
  const wrapper = shallow(
    <AddExpensePage onSubmit={onSubmit} history={history} />
  );
  wrapper
    .find('ExpenseForm')
    .simulate('submit', expenses[1]);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(onSubmit).toHaveBeenCalledWith(expenses[1]);
});