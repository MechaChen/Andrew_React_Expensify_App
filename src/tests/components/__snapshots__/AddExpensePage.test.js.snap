// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddExpesePage 1`] = `
<div>
  <h1>
    Add Expense
  </h1>
  <ExpenseForm
    expense={
      Object {
        "amount": "",
        "createdAt": "1970-01-01T00:00:00.000Z",
        "description": "",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
</div>
`;
